app/src/main/java/com/github/swent/echo/connectivity/NetworkServiceImpl.kt

```diff
@@ -0,0 +1,40 @@
+package com.github.swent.echo.connectivity
+
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
+class NetworkServiceImpl(connectivityManager: ConnectivityManager) : NetworkService {
+    private val _isOnline = MutableStateFlow(false)
+
+    init {
+        val caps = connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
+        _isOnline.value = caps?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) ?: false
+
+        val request =
+            NetworkRequest.Builder()
+                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                .build()
+
+        val networkCallback =
+            object : ConnectivityManager.NetworkCallback() {
+                override fun onAvailable(network: Network) {
+                    _isOnline.value = true
```

Comment:
Knitpick: What do you think about updating the _isOnline.value asynchronously in the main thread using CoroutineScope(Dispatchers.Main).launch { _isOnline.value = true }? Doing this would prompt the UI quickly as soon as the change in network has been detected. What do you think?

Reply:
As this is a very in-expensive operation (just set a value) I don't think we need to launch a coroutine for that. As launching a coroutine requires some work, idk if it really would be quicker...